csrank/objectranking/lambdamart.py

```diff
@@ -0,0 +1,370 @@
+            
+            pool = Pool()
+            for lambda_val, w_val, query_key in pool.map(compute_lambda_weights, zip(true_scores, pred_scores, sorted_query_pair, idcg, query_keys), chunksize=1):
+                indexes = query_indexes[query_key]
+                lambdas[indexes] = lambda_val
+                w[indexes] = w_val
+            pool.close()
+            # filename = str("lmartDebug2_"+str(k)+"_lambdas.txt")
+            # np.savetxt(filename, lambdas)
+
+            #Hyperparameters to be considered are :"min_samples_split","max_depth","min_samples_leaf","max_leaf_nodes"
+            tree = DecisionTreeRegressor(criterion="mse",
+                                         splitter="best",
+                                         max_depth=self.max_depth,
+                                         min_samples_split=self.min_samples_split,
+                                         min_samples_leaf=self.min_samples_leaf,
+                                         min_weight_fraction_leaf=0.,
+                                         max_features=None,
+                                         random_state=9,
+                                         max_leaf_nodes=self.max_leaf_nodes,
+                                         min_impurity_decrease=0.,
+                                         min_impurity_split=None)
+            tree.fit(training_data[:,2:], lambdas)
+            self.trees.append(tree)
+            prediction = tree.predict(training_data[:,2:])
+            predicted_scores += prediction * self.learning_rate
+    
+    def _predict_scores_fixed(self, X, **kwargs):
+        n_instances, n_objects, n_features = X.shape
+        self.logger.info("For Test instances {} objects {} features {}".format(*X.shape))
+        X1 = X.reshape(n_instances * n_objects, n_features)
+        scores = np.zeros(n_instances * n_objects)
+        for tree in self.trees:
+            scores += tree.predict(X1)
+        scores = scores.reshape(n_instances, n_objects)
+        return scores
+
+    def predict_scores(self, X, **kwargs):
+        return super().predict_scores(X, **kwargs)
+
+    def predict_for_scores(self, scores, **kwargs):
+        return ObjectRanker.predict_for_scores(self, scores, **kwargs)
+
+    def predict(self, X, **kwargs):
+        return super().predict(X, **kwargs)
+    
+    def _predict(self, data):
+        """
+            Predict the
+        """        
+        data = np.array(data)
+        query_indexes = self._group_queries(data, 0)
+        predicted_scores = np.zeros(len(data))
+        for query in query_indexes:
+            results = np.zeros(len(query_indexes[query]))
+            for tree in self.trees:
+                results += self.learning_rate * tree.predict(data[query_indexes[query], 1:])
+            predicted_scores[query_indexes[query]] = results
+        return predicted_scores
+
+    def _validate(self, data, k=10):
```

Comment:
Docstring

The model should only be validated from the outside. This method can be removed.